subiquity/server/snapd/types.py

```diff
@@ -205,16 +205,67 @@ class StorageEncryption:
 
 
 @snapdtype
-class SystemDetails:
+class AvailableOptional:
+    snaps: List[str] = attr.Factory(list)
+    components: Dict[str, List[str]] = attr.Factory(dict)
+
+
+class ModelSnapType(enum.Enum):
+    KERNEL = "kernel"
+    GADGET = "gadget"
+    BASE = "base"
+    SNAPD = "snapd"
+    APP = "app"
+
+
+class PresenceValue(enum.Enum):
+    REQUIRED = "required"
+    OPTIONAL = "optional"
+
+
+@snapdtype
+class Presence:
+    presence: PresenceValue
+
+
+@snapdtype
+class ModelSnap:
+    default_channel: str
+    id: str
+    name: str
+    type: NonExhaustive[ModelSnapType]
+    components: Optional[Dict[str, Presence]] = None
+
+
+@snapdtype
+class Model:
+    architecture: str
+    snaps: List[ModelSnap]
+
+    def snap_of_type(self, typ: ModelSnapType) -> Optional[ModelSnap]:
+        for snap in self.snaps:
+            if snap.type == typ:
+                return snap
+        return None
```

Comment:
nit: this only gets the first of the type in the list. From the name and usage it seems expected that there's only one of this type, which is true for the kernel (could we ever have more?), but it wouldn't work for say `typ=ModelSnapType.APP` where we already have multiple.

I think instead we want something like:
```suggestion
    def snaps_of_type(self, typ: ModelSnapType) -> List[ModelSnap]:
        return [snap for snap in self.snaps if snap.type == typ]
```
and have the caller make the assertion about the number of them?

I wouldn't say it's blocking and think it's fine to fix this in a follow up though.

Reply:
https://github.com/canonical/subiquity/pull/2168/commits/42ca09c59ea3d5c0b3aa620f76d966ff6e8e0139